component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -336,6 +327,135 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
+    }
+
+    /**
+     * Checks the backup codes for email otp.
+     *
+     * @param context           AuthenticationContext.
+     * @param userToken         EmailOTP token.
+     * @param authenticatedUser AuthenticatedUser
+     * @throws AuthenticationFailedException
+     */
+    private void checkWithBackUpCodes(AuthenticationContext context, String userToken,
+                                      AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
+
+        String[] savedOTPs = null;
+        String username = authenticatedUser.getUserName();
+        String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+        UserRealm userRealm = getUserRealm(username);
+        try {
+            if (userRealm != null) {
+                UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+                if (userStoreManager != null) {
+                    String savedOTPString = userStoreManager
+                            .getUserClaimValue(tenantAwareUsername,
+                                    EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM, null);
+                    if (StringUtils.isNotEmpty(savedOTPString)) {
+                        savedOTPs = savedOTPString.split(",");
+                    }
+                }
+            }
+            // Check whether there is any backup OTPs and return.
+            if (ArrayUtils.isEmpty(savedOTPs)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("The claim " + EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM + " does " +
+                            "not contain any values.");
+                }
+                throw new AuthenticationFailedException("The claim " +
+                        EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM +
+                        " does not contain any values", authenticatedUser);
+            }
+            if (isBackUpCodeValid(savedOTPs, userToken)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Found saved backup Email OTP for user :" + authenticatedUser);
+                }
+                context.setSubject(authenticatedUser);
+                savedOTPs = (String[]) ArrayUtils.removeElement(savedOTPs, userToken);
+                userRealm.getUserStoreManager().setUserClaimValue(tenantAwareUsername,
+                        EmailOTPAuthenticatorConstants.OTP_BACKUP_CODES_CLAIM, String.join(",", savedOTPs),
+                        null);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("User entered OTP :" + userToken + " does not match with any of the saved " +
+                            "backup codes.");
+                }
+                context.setProperty(EmailOTPAuthenticatorConstants.CODE_MISMATCH, true);
+                throw new AuthenticationFailedException("Verification Error due to Code " + userToken + " " +
```

Comment:
Shall we change this to "Verification error due to code " + userToken+ " mismatch."